pil-analyzer/src/traits_resolver.rs

```diff
@@ -60,26 +91,176 @@ impl<'a> TraitsResolver<'a> {
+                                        ));
+                                    }
+                                }
+                            }
+                        }
+                        None => {
+                            continue;
+                        }
+                    }
+                }
+            }
+        }
+        solved_impls
+    }
+
+    /// Creates a dictionary that associates each child reference with a generic type along with the references that called it
+    /// From this dictionary, it resolves the parent/child relationships to obtain the value of the generic type
+    fn build_reference_path(
+        definitions: &Vec<(&String, &Expression)>,
+        defined_traits: &HashSet<&String>,
+    ) -> HashMap<String, HashSet<Vec<Type>>> {
+        let mut result = HashMap::new();
+
+        // Compute all pairs of references (name, type) for each child and store them associated with the parent's name
+        // ["Child": [
+        //      ("Parent1", [Some(type_args)]),
+        //      ("Parent2", [Some(type_args)])
+        // ],...]
+        for (symbol, expr) in definitions {
+            let mut references = Vec::new();
+            expr.all_children().for_each(|e| {
+                if let Expression::Reference(
+                    _,
+                    Reference::Poly(PolynomialReference {
+                        name,
+                        type_args: Some(type_args),
+                        ..
+                    }),
+                ) = e
+                {
+                    if (!type_args.is_empty()) && (name != *symbol) {
+                        references.push((name, type_args));
+                    }
+                }
+            });
+            for (name, type_args) in references {
+                result
+                    .entry(name)
+                    .or_insert_with(Vec::new)
+                    .push((*symbol, type_args));
+            }
+        }
+
+        Self::combine_type_vars_paths(&result, defined_traits)
+    }
+
+    /// Solve the initial asociation and builds a dictionary that maps the
+    /// names of traits containing generic types to their represented types.
+    fn combine_type_vars_paths(
+        input: &HashMap<&String, Vec<(&String, &Vec<Type>)>>,
+        defined_traits: &HashSet<&String>,
+    ) -> HashMap<String, HashSet<Vec<Type>>> {
+        let mut result: HashMap<String, HashSet<Vec<Type>>> = HashMap::new();
+
+        for t in defined_traits {
+            match input.get(t) {
+                Some(children) => {
+                    for (father, type_arg) in children {
+                        if type_arg.iter().any(|ty| matches!(ty, Type::TypeVar(_))) {
+                            let new_type = Self::solve_type_vars(input, father, type_arg);
+                            result.insert(t.to_string(), new_type);
+                        }
+                    }
+                }
+                None => continue,
```

Comment:
Is this reachable because the keys of `input` are just the traits which contain generic types here? Related to the comment above, having all traits in both would simplify the code here as the `Some` case seems to work fine for empty vectors